### PR TITLE
Feat pyodide kernel

### PR DESCRIPTION
**Pull request by stefnotch**
This moves Pyodide off the main thread to run in a web worker.

- Run Pyodide in web worker
  - I chose to not inline this web worker, since I'm not sure if it's a good idea to inline a 40+ KB large file.
  - The worker has a separate rollup entrypoint, so that it gets exactly the code that it needs. Importing things into a web worker sadly isn't supported in all browsers yet.
  - This uses the `Atomics.wait` approach for pausing the web worker https://github.com/pyodide/pyodide/issues/1504#issuecomment-882105354
  - Here is an explanation of how the entire [data passing/web worker accessing objects from the main thread](https://github.com/pyodide/pyodide/issues/1504#issuecomment-885136725) works
- Adds a generic kernel manager
- Keep public API pretty much the same
- Updated dependencies
- Reformatted files with Prettier
- Added option to run Pyodide in main thread again
- Flex on Jupyterlite with `print(int(input()) * 2)`. Also note how the behaviour of `input()` has been fixed

### CODE REVIEW

**Comment by stefnotch** [comment]
The biggest remaining issue is passing data from the main thread to the worker and back. To truly work well, this will need a pretty fancy solution along the lines of

1. Serialize small things that can be serialized
2. Wrap the other things in comlink
3. Hand them over to Python
4. When some property is accessed, the stuff behind it will be fetched/requested synchronously

I'll also need to implement a synchronous way of "getting a variable" to make sure that things like the following still work, without first having to send over every single variable

```py
from js import window
```
Relevant https://github.com/pyodide/pyodide/pull/1146

Relevant https://github.com/pyodide/pyodide/issues/1504#issuecomment-882315144

**Review by gzuidhof — COMMENTED** [review]
Four high level pieces (we can push some into a future PR, especially the testing one)
* Let's still support the existing workflow (and perhaps even keep the current function signatures for them).
* I think we can wrap the `kernel` or whatever into some stateful `class`. Whether it should be one class, or two (one for async with webworker, one for main thread) I don't know - I think you'll have a better feeling for this once you try it. It would be nice if this also had say a method to "preload" pyodide (e.g. for Illumidesk we know they will run Python sooner or later, we might as well load it before they run the first cell).
* The main thread basecase should still work in Safari and friends. Ideally the webworker-mode would also work for anything except for "syncified" stuff (`input`).
* It would be really great to be able to test this with at least some smoke tests. Ideally automated (feel free to pick the most suitable test framework, I imagine we will need a real (headless) browser), but at the very least a demo html page with some sanity checks (e.g. running matplotlib).

File: src/global.ts
Comment (gzuidhof) [inline src/global.ts]: If we never expect this to happen, maybe let's console error or warn here? Something like:

console.warn("Pyodide webworker received unexpected message:", e)

File: src/global.ts
Comment (gzuidhof) [inline src/global.ts]: Maybe `console.warn` in these cases?

File: src/global.ts
Comment (gzuidhof) [inline src/global.ts]: If we go for a stateful `Kernel` or whatever call it, perhaps this could be a function you can pass in (with this as default)

File: src/worker/emscripten-utils.ts
Comment (gzuidhof) [inline src/worker/emscripten-utils.ts]: I think we can depend on `TextDecoder` being present and don't need this safety, I think every browser that supports WASM has it. I guess then we can delete a bunch of code here? I believe there's also a `TextEncoder`?

**Review by gzuidhof — APPROVED** [review]

File: src/worker/emscripten-utils.ts
Comment (stefnotch) [reply src/worker/emscripten-utils.ts]: Yes, we definitely can. Then again, that part of the code is mostly glue-code that's copied straight from emscripten. I'm hoping that someday Pyodide will support that out of the box and then we'll be able to get rid of it.

File: src/global.ts
Comment (stefnotch) [reply src/global.ts]: Agreed. This is definitely one of those functions that would be useful for more programming languages/cell types

**Comment by stefnotch** [comment]
> Let's still support the existing workflow (and perhaps even keep the current function signatures for them).

Agreed, let's still support running Pyodide on the main thread. This form of Pyodide wouldn't support remote filesystem access though, since the Python filesystem API is one of those APIs that expect things to happen synchronously. And that's something we can only pull off by pausing a thread (or by using one of the other solutions https://github.com/pyodide/pyodide/issues/1504#issuecomment-882105354 ). I suppose that'd be an acceptable limitation.


> I think we can wrap the `kernel` or whatever into some stateful `class`. Whether it should be one class, or two (one for async with webworker, one for main thread) I don't know - I think you'll have a better feeling for this once you try it.

Okay, I'll try this out. That's an important step towards enabling the whole kernels stuff anyways.

>  It would be nice if this also had say a method to "preload" pyodide (e.g. for Illumidesk we know they will run Python sooner or later, we might as well load it before they run the first cell).

For preloading, there is also https://github.com/pyodide/pyodide/issues/1576

> The main thread basecase should still work in Safari and friends. Ideally the webworker-mode would also work for anything except for "syncified" stuff 

I believe that it does, but I don't have a Safari browser at hand to test this with.

**Comment by stefnotch** [comment]
Tasks (for myself):
- [x] Finish the AsyncMemory redesign/rewrite
- [x] Serialize object to AsyncMemory
- [x] Deserialize object from AsyncMemory
- [x] Finish the data passing, depends on ^
- [x] Finish refactoring the worker code into a class with a well defined API

and maybe also
- [x] Re-enable running Pyodide on the main thread

**Comment by stefnotch** [comment]
Because I just saw the issue on the Pyodide issue tracker: 
- [x] Test if this works with `from js import whateverClipboardApiThereIs`
https://github.com/pyodide/pyodide/issues/1735

**Comment by stefnotch** [comment]
Proxies have one irritating special case: Functions. 
Javascript functions have a `this` context, which will be wrong when one naively returns the function from the proxy. One workaround is wrapping the function into another proxy and changing the `this` context accordingly. Which mostly works, but breaks down in the case of nested proxies. 

If you're aware of a decent solution for this, I'd love to hear it. For now, I'll probably just implement a 'is proxy' check...

```js
exclude = new Set()


x = new Proxy(globalThis, {
 get(target, prop, receiver) {
        if (typeof prop === "string" && exclude.has(prop)) {
          target = globalThis;
        }
		console.log("x-proxy get");
        const value = Reflect.get(target, prop, receiver);
		console.log("x-proxy get returned", { value, target, prop, receiver });
        if (typeof value !== "function") return value;
		
		console.log("x-proxy new function proxy")
        return new Proxy(value, {
          apply(_, thisArg, args) {
			console.log("x-proxy function called")
            const calledWithProxy = thisArg === receiver; // <-- this check is meh
			console.log("x-proxy function called and", calledWithProxy)
            return Reflect.apply(value, calledWithProxy ? target : thisArg, args);
          },
        });
      }
})


y = new Proxy(x, {
 get(target, prop, receiver) {
        if (typeof prop === "string" && exclude.has(prop)) {
          target = globalThis;
        }

		console.log("y-proxy get");
        const value = Reflect.get(target, prop, receiver);
		console.log("y-proxy get returned", { value, target, prop, receiver });
        if (typeof value !== "function") return value;
		
		console.log("y-proxy new function proxy")
        return new Proxy(value, {
          apply(_, thisArg, args) {
			console.log("y-proxy function called")
            const calledWithProxy = thisArg === receiver; // <-- this check is meh
			console.log("y-proxy function called and", calledWithProxy)
            return Reflect.apply(value, calledWithProxy ? target : thisArg, args);
          },
        });
      },
	  has() {
		return false;
	  }
})

y.setInterval(() => {}, 1) // <-- this ends up failing
```

**Comment by stefnotch** [comment]
For matplotlib, it might be worthwhile to use a different patch. I'm still working on the proxy-ing, but it seems like that'll end up causing hundreds of calls to the main thread.
https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/py/pyolite/pyolite/patches.py

And implementing a truly good proxy that can also delay function execution is slightly out of scope for now.

**Comment by stefnotch** [comment]
The changes required to run matplotlib *over the data passing proxy* are here 
https://github.com/stefnotch/starboard-python/tree/feat-illegal-stuff


image

**Comment by stefnotch** [comment]
Remaining pieces of work
- ~~Running this in the main thread again~~
- ~~matplotlib support~~

**Comment by stefnotch** [comment]
> 
> 
> Because I just saw the issue on the Pyodide issue tracker:
> 
>     * [x]  Test if this works with `from js import whateverClipboardApiThereIs`
>       [ModuleNotFoundError: No module named 'ctypes' pyodide/pyodide#1735](https://github.com/pyodide/pyodide/issues/1735)

Yes, it mostly works.
```py
from js import navigator

navigator.clipboard.writeText("Cats are cute")
print("Indeed")
```
This works, however, you cannot await the promise. If you do, it'll result in an error and the kernel will hang. I guess it's good enough for now.

File: package.json
Comment (gzuidhof) [inline package.json]: Maybe it's bit too specific to put it here (i.e. for me it's `Edge Dev` instead of `Edge`), there are two alternatives:

1. You can invoke the test using
```
npm run test -- --browser_exec_path <my path>
```

Or create the `.starlit.yml` file in the current dir or your home dir.

File: src/global.ts
Comment (gzuidhof) [inline src/global.ts:186]: This used to have `"https://cdn.jsdelivr.net/pyodide/v0.17.0/full/"` as fallback, do you think we should still include it here as final resort?

File: src/main-thread-pyodide.ts
Comment (gzuidhof) [inline src/main-thread-pyodide.ts]: nit
```suggestion
  const pyodideWorkerOptions = opts.options as PyodideWorkerOptions;
```

File: src/main-thread-pyodide.ts
Comment (gzuidhof) [inline src/main-thread-pyodide.ts:30]: It's great that you wrapped it in the same interface! 👍

File: src/util.ts
Comment (gzuidhof) [inline src/util.ts:1]: Nice :). I've read about this but never used it, I guess I should put this at the end of every switch-like statement.

File: test/python2.nb
Comment (gzuidhof) [inline test/python2.nb:1]: Woo :)

**Review by gzuidhof — APPROVED** [review]
Thank you, this is ready to merge when you feel comfortable :). I have added you as a collaborator, and as a maintainer on NPM. So you can mint new versions on NPM.

Maybe let's add a small changelog somewhere?

File: src/global.ts
Comment (stefnotch) [reply src/global.ts:186]: The `pyodide-worker.ts` has it as a fallback. Should I move the fallback to over here? Or have it in both places?

File: package.json
Comment (stefnotch) [reply package.json]: Good point

File: src/util.ts
Comment (stefnotch) [reply src/util.ts:1]: Yeah, it's really useful! I'm now copying that assertUnreachable into every single one of my projects 😆

File: src/global.ts
Comment (gzuidhof) [reply src/global.ts:186]: Ah, I didn't notice, wherever you think it makes best sense :) No strong opinion

**Pull request merged**
